File: samples/client/3_1_0_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/TypeValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.openapijsonschematools.client.exceptions.ValidationException;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.List;
import java.util.Map;
import java.util.Set;

public class TypeValidator implements KeywordValidator {
    public final Set<Class<?>> type;

    public TypeValidator(Set<Class<?>> type) {
        this.type = type;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        ValidationData data
    ) {
        Class<?> argClass;
        if (data.arg() == null) {
            argClass = Void.class;
        } else if (data.arg() instanceof List) {
            argClass = List.class;
        } else if (data.arg() instanceof Map) {
            argClass = Map.class;
        } else {
            argClass = data.arg().getClass();
        }
        if (!type.contains(argClass)) {
            throw new ValidationException("invalid type");
        }
        return null;
    }
}

File: samples/client/3_1_0_unit_test/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/TypeValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class TypeValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        if (data.schema.type == null) {
            return null
        }
        val argClass = when (data.arg) {
            null -> {
                Void::class.java
            }
            is List<*> -> {
                List::class.java
            }

            is Map<*, *> -> {
                Map::class.java
            }

            is Boolean -> {
                Boolean::class.java
            }

            is Int -> {
                Int::class.java
            }
            is Long -> {
                Long::class.java
            }
            is Double -> {
                Double::class.java
            }
            is Float -> {
                Float::class.java
            }

            else -> {
                data.arg.javaClass
            }
        }
        if (!data.schema.type.contains(argClass)) {
            throw ValidationException("invalid type")
        }
        return null
    }
}